### PR TITLE
Improve "weird" transitive package dependencies of Gardener binaries

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity
/kind technical-debt

**What this PR does / why we need it**:
This PR improves "weird" transitive package dependencies for `gardener-controller-manager` (GCM) by

- moving garden-specific and shoot-specific utilities from `pkg/operation/shoot` to `pkg/utils/gardener`
- moving GCM-specific code to `app/bootstrappers`
- moving the `extensions.Id` function to `gardenerutils.ExtensionsID`

As a result, the GCM binary is no longer dependent on the majority of code that actually belongs to `gardenlet` only.

I also checked the dependencies of other binaries, but I think they are mostly fine, and I don't see any immediate necessity to invest further/change something.

**Which issue(s) this PR fixes**:
Part of #6016

**Special notes for your reviewer**:
See also https://github.com/gardener/gardener/pull/6781.

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/utils/gardener/shoot.go [hunk 1]
```diff
@@ -514,3 +516,169 @@ func ExtractSystemComponentsTolerations(workers []gardencorev1beta1.Worker) []co
...
+	return "unable to figure out which secret should be used for dns"
+}
+
+// IsIncompleteDNSConfigError returns true if the error indicates that not the DNS config is incomplete.
+func IsIncompleteDNSConfigError(err error) bool {
+	_, ok := err.(*IncompleteDNSConfigError)
+	return ok
+}
+
+// ConstructInternalClusterDomain constructs the internal base domain pof this shoot cluster.
```
Comment (timuthy) [inline pkg/utils/gardener/shoot.go]:
```suggestion
// ConstructInternalClusterDomain constructs the internal base domain for this shoot cluster.
```

File: pkg/apis/core/validation/controllerregistration.go [hunk 2]
```diff
@@ -91,7 +90,7 @@ func ValidateControllerRegistrationSpec(spec *core.ControllerRegistrationSpec, f
 			allErrs = append(allErrs, field.Required(idxPath.Child("type"), "field is required"))
 		}
 		if t, ok := resources[resource.Kind]; ok && t == resource.Type {
-			allErrs = append(allErrs, field.Duplicate(idxPath, extensions.Id(resource.Kind, resource.Type)))
+			allErrs = append(allErrs, field.Duplicate(idxPath, fmt.Sprintf("%s/%s", resource.Kind, resource.Type)))
```
Comment (timuthy) [inline pkg/apis/core/validation/controllerregistration.go]: Why is `gardenerutils.ExtensionsID` not used here? The validation package already imports `github.com/gardener/gardener/pkg/utils/gardener`.

File: pkg/utils/gardener/garden.go [hunk 3]
```diff
@@ -28,9 +28,34 @@ import (
 	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
 	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
 	"github.com/gardener/gardener/pkg/utils"
-	gardenerutils "github.com/gardener/gardener/pkg/utils/gardener"
 )
 
+// Builder is an object that builds Garden objects.
+type Builder struct {
```
Comment (timuthy) [inline pkg/utils/gardener/garden.go]: Isn't this builder still very `operation` specific? I'm not quite sure why we want to move this out.

Comment (rfranzke) [reply pkg/utils/gardener/garden.go]: Typically, the builder code is maintained close to the built object, isn't it? I have no preference - we can also move it elsewhere (what do you suggest). Maybe we can also eliminate it entirely and directly construct the `Garden` object?

**Review by timuthy — COMMENTED** [review]
Thank you!
/lgtm

**Review by acumino — COMMENTED** [review]
/approve
/needs-rebase

**Review by timuthy — COMMENTED** [review]
/lgtm
Thank you!

**Pull request merged**